### PR TITLE
[pub sub] - Event subscription using event filter

### PR DESCRIPTION
**Pull request by patrickkuo**
* replaced `subscribeMoveEventsByType` with `subscribeEvent`
* subscription now support EventFilter input, for example, to subscribe to all publish event, `sui_subscribeEvent [{"EventType":"Publish"}]`
* Added sender and tx digest fields to `Event::MoveEvent`, `Event::Publish`, `Event::RemoveObject` and `Event::NewObject` 
* Added function info to `Event::NewObject` and `Event::MoveEvent`

Todos:
- [x] Docs
- [x] Unit test

### CODE REVIEW

**Comment by patrickkuo** [comment]
> Hey @patrickkuo thanks for adding stuff like package ID and module name and function name to the other events. What is the sender btw? Is it who created the transaction?
> 
> My first thought is, since many events will all share common details like package ID, module name, function name, etc., I wonder if we should just move it to `EventEnvelope`, so that we can reduce the clutter in each event?
> 
> I'll spend more time to review everything soon.

I will explore that, some attributes are from the adaptor (function name, module name etc) I am not sure if we store those and if they can be retrieved when we create the Envelope in eventHandler.

**Comment by patrickkuo** [comment]
@Clay-Mysten I have added new docs for event pubsub in this PR, I have put them in a new file for now, please feel free to change/ move it around if you think it belongs somewhere else :)

**Comment by Clay-Mysten** [comment]
> @Clay-Mysten I have added new docs for event pubsub in this PR, I have put them in a new file for now, please feel free to change/ move it around if you think it belongs somewhere else :)

Thanks, Patrick. I am reviewing and editing as a direct commit here, if needed. I will then add the file to navigation and help you announce and link into it from the usual places. Thanks for including me!

**Comment by Clay-Mysten** [comment]
Great work, Patrick. I made my edits directly. When this is in and you are ready, let's add the page to navigation (between Use Sui RPC Server and Employ SuiJSON Format?) and announce. Please advise. Thanks!

File: crates/sui-types/src/event.rs
Comment (tnowacki) [inline crates/sui-types/src/event.rs:85]: For these module/function pairs. Could you make it something like `transaction_module` and `transaction_function`?
Or something similar.

My concern is it is not obvious what these are doing here. At first I thought these were somehow related to the `type_`, not the transaction that created the event.

File: crates/sui-types/src/event_filter.rs
Comment (velvia) [inline crates/sui-types/src/event_filter.rs:36]: Really really good.

**Review by velvia — CHANGES_REQUESTED** [review]
Ok, I think I've left enough comments for now.  Great great work @patrickkuo I especially love the generic event filter mechanism.   Its really nice.    We can add it to the event store no problem.  

Sender can be moved to EventEnvelope because the sender is available from a certificate.  I can in fact add that as part of my other PR that's coming.

File: crates/sui-types/src/event.rs
Comment (tnowacki) [reply crates/sui-types/src/event.rs:84]: There is no string interning for `Identifier`, its just validation

File: crates/sui-types/src/event.rs
Comment (patrickkuo) [reply crates/sui-types/src/event.rs:109]: I think we can keep both for now, there are talk in another thread about adding restriction in move to restrict EventType to be originated from the same package and module, we can remove package/module when that happened.

For now should I just leave it as it is? We can remove them later for space saving/optimisation later.

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs:85]: Oh, I see.  @patrickkuo I wonder if we should switch this to just String, or FlexStr (which can accommodate `&str` as well)

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs:110]: Per conversation from yesterday it sounds like we now want to get rid of the "function" for clarity?

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs]: Ah.... "initiator" would have been fine too, but we should just stick with the term used elsewhere eg "sender".
BTW "instigator" has a negative meaning, like someone who started a bad thing. :)

Comment (patrickkuo) [reply crates/sui-types/src/event.rs]: Oh ...  😅

File: crates/sui-types/src/event.rs
Comment (longbowlu) [reply crates/sui-types/src/event.rs:110]: PS2: I just realize that the thing we return to client `SuiEvent` is not authenticatable. To make it so, we need to at least return a TransactionDigest so clients can verify by pulling the txn by themselves. 

I will add that based on this PR .

File: crates/sui-types/src/event_filter.rs
Comment (longbowlu) [inline crates/sui-types/src/event_filter.rs:50]: -> `object_id` to make it consistent with line 59?

File: doc/src/build/pubsub.md
Comment (longbowlu) [inline doc/src/build/pubsub.md:2]: awesome! cc  @Clay-Mysten

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs:110]: @longbowlu @patrickkuo ok, there are some differences between the EventStore and subscription module that we should discuss, related to this:
1. In EventStore, we store package_id and module as separate columns, to make the querying more flexible (let's say user queries only by package ID).   I think it's good to leave them as separate for flexibility.   In the future they would also compress better if left separate.
2. From an event subscription perspective it makes very little difference if package_id and module are in EventEnvelope or Event, they are 1:1.  However, it makes a big difference if we include them in Event for TransactionEffects, because including them means that the package id and module are repeated in each event, and that bloats the `TransactionEffects`.   

We can leave them here for now, but there are more compact ways to represent, like if each event had a reference to the source TransactionData number within the Transaction batch, and through that we can get the source MOVE package ID and module.

File: crates/sui-json-rpc-api/src/rpc_types.rs
Comment (patrickkuo) [reply crates/sui-json-rpc-api/src/rpc_types.rs:1463]: ya agree!, I will remove it

File: doc/src/build/pubsub.md
Comment (Clay-Mysten) [reply doc/src/build/pubsub.md:2]: > awesome! cc @Clay-Mysten

Thanks! Re-reviewing/re-editing now.

**Review by Clay-Mysten — APPROVED** [review]
The new pubsub.md file LGTM after a second round of edits. Thanks, Patrick!

File: crates/sui-types/src/event.rs
Comment (longbowlu) [reply crates/sui-types/src/event.rs:108]: @velvia  I re-read your to comment and have this q: " events all appear in TransactionEffects in a Vec, and we do not know which source SingleTransaction produced the event." A `TransactinoEffects` is only associated with a `Transaction` so there shouldn't be difficulties in linking back `Events` back to `Transaction` from that perspective.

after some pondering, I think I'm sold that we should put `package/module` into `EventEnvelope` instead of `Event`, reasons being:
1. the data bloat problem @velvia  pointed out above
2. most Events enums have `package` and `function`, so it's a good idea to put it into a wrapper struct ?
3. currently we seem to the `Event` enums, (except`packge/function`), they depict "what happened", but not "how/when/where it happened". For example we don't put timestamp (where) in `Event`. Following this logic, `sender` probably does not belong to `Event` either, except `TransferObject`.

I'm OK with leaving this PR as-is but let's sync on what to put in each `Event` enum next week

**Review by longbowlu — APPROVED** [review]

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs:108]: Yes, `sender` should belong in the envelope as well since there is only one Sender per TX.  

Not sure what is meant about a wrapper struct

File: crates/sui-json-rpc-api/src/rpc_types.rs
Comment (patrickkuo) [reply crates/sui-json-rpc-api/src/rpc_types.rs:1200]: I think we should, and I think we can use a const TransactionDigest for system events, so we don't have to deal with Options.

Comment (velvia) [reply crates/sui-json-rpc-api/src/rpc_types.rs:1200]: I'm not sure about that.  We should check with others including @lxfind and @gdanezis but here's my take on it:

I don't think we should manufacture TX digests for events which aren't tied to transactions, because TX digests have a specific meaning: they refer to valid transactions which are certified.  If there could be TX digests which are "fake" then we need to explain this, and have special cases in our code to deal with it.   Customers must also deal with this.  If a customer sees a TX digest for a "checkpoint event", how will they interpret it?  Will they think, gee this must point to the last TX before the checkpoint?   They might then try to use our APIs to retrieve such a transaction, only to find there is no such transaction.   This seems like a bad idea.

Maybe, if that digest was all zeroes, then people might think gee this is something different, but I'm not sure that makes it better.

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs:85]: @patrickkuo I think what you're saying is this:

- We can move the fields to EventEnvelope
- However we only populate Events in adapter.rs
- So populating EventEnvelope in event_handler is not unambiguously possible right now, unless we add something else to an event

I think it's fine to leave it for now.

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:71]: Couldn't we put all this code into `to_move_struct_with_resolver` or the equivalent in MoveObject?  Just wondering if we can avoid duplicating all this code

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:92]: Again I'm wondering what is the purpose of creating this new function vs the existing stuff to create json values

File: crates/sui-json-rpc-api/src/rpc_types.rs
Comment (velvia) [inline crates/sui-json-rpc-api/src/rpc_types.rs:1454]: Isn't this redundant?  `MoveEventType` == `All(vec![Package(..), Module(...)])`

**Review by velvia — APPROVED** [review]
Just a couple more comments. 

I think `sender` could be moved out of Event into `EventEnvelope` for sure, because sender is present in every `Transaction` already.   However if you want to do log a ticket and we can do this in the future along with moving package and module away from events, that's fine too.

File: crates/sui-json-rpc-api/src/rpc_types.rs
Comment (patrickkuo) [reply crates/sui-json-rpc-api/src/rpc_types.rs:1454]: this will be redundant when the only allow event from the same module restriction is enforced in Move code, we can remove the usage of Package and Module filter for MoveEvents.

**Comment by Clay-Mysten** [comment]
Thanks, Patrick. I will add the new PubSub page to navigation and then announce if alright with you. Thanks!

**Comment by patrickkuo** [comment]
> Thanks, Patrick. I will add the new PubSub page to navigation and then announce if alright with you. Thanks!

Thanks Clay! Regarding announcements maybe wait for event query api? So we can have one announcement for all new Event functionalities?

**Comment by Clay-Mysten** [comment]
> > Thanks, Patrick. I will add the new PubSub page to navigation and then announce if alright with you. Thanks!
> 
> Thanks Clay! Regarding announcements maybe wait for event query api? So we can have one announcement for all new Event functionalities?

Thanks, Patrick! I went ahead and started the process in these PRs:
https://github.com/MystenLabs/sui/pull/2774
https://github.com/MystenLabs/sui/pull/2775
https://github.com/MystenLabs/sui/pull/2777

We can always hold those for now. Please just let me know.

**Comment by Clay-Mysten** [comment]
@patrickkuo , all we need is approval of the navigation PR to make this content live:
https://github.com/MystenLabs/sui/pull/2836

Please take a look and let me know if you have questions. Thanks!

**Pull request merged**
